eventmesh/LockManager.js

```diff
@@ -0,0 +1,95 @@
+'use strict';
+
+const _ = require('lodash');
+const config = require('../common/config');
+const CONST = require('./constants');
+
+const {
+  Etcd3
+} = require('etcd3');
+const client = new Etcd3({
+  hosts: config.etcd.url,
+  credentials: {
+    rootCertificate: Buffer.from(config.etcd.ssl.ca, 'utf8'),
+    privateKey: Buffer.from(config.etcd.ssl.key, 'utf8'),
+    certChain: Buffer.from(config.etcd.ssl.crt, 'utf8')
+  }
+});
+
+class LockManager {
+  /*
+  Lock is tracked via the key value "resource/lock/details". 
+  The value is a JSON type and the structure looks like the following.
+  {count: INT, operaitonType:STRING}
+  when a lock is taken, count value is set to 1.
+  operationType value is either "READ or "WRITE", depending on if the ongoing operation will make sync operations like bind/unbind wait for it.
+
+  This value is read and updated to acquire lock. In order to make this two different operations synchronous, we do it taking a lock on the "resource/lock" key, which would ensure synchrounous execution of the read and update operation.
+
+  Lock Algorithm looks like the following
+      1. Lock "resource/lock"
+      2. Check "resource/lock/details" value
+      3. if lock value > 1 then lock is already acquired by someone, 
+         unlock "resource/lock" and throw exception.
+      4. else, set the lock details as 
+         {count: 1, operationType: "READ"} 
+      5. release the lock on "resource/lock"
+  */
+  lock(resource, operationType) {
+    const lock = client.lock(resource + CONST.LOCK_KEY_SUFFIX);
+    return lock.ttl(CONST.LOCK_TTL).acquire()
+      .then(() => {
+        return client.get(resource + CONST.LOCK_DETAILS_SUFFIX).json();
+      })
+      .then(lockDetails => {
+        if (_.get(lockDetails, 'count') > 0) {
+          return lock.release().then(() => {
+            throw Error('Could not acquire lock for ' + resource);
+          });
+        } else {
+          const newLockDetails = {};
+          newLockDetails.count = 1;
+          newLockDetails.operationType = operationType;
+          return client.put(resource + CONST.LOCK_DETAILS_SUFFIX).value(JSON.stringify(newLockDetails));
+        }
+      })
+      .then(() => lock.release())
+      .catch(e => {
+        this.unlock(resource);
```

Comment:
@subhankarc I think, we cannot simply call unlock if any exception occurs. We want to unlock when we have put the count = 1 and an exception occurred later (particularly when 2nd lock.release() fails). If lets say, acquire() itself failed or first lock.release() failed, we will be blindly unlocking, which might override count set by other operation.
We can keep a boolean flag which is set after client.put() and only if this flag is set, we unlock. Does this make sense?

Reply:
Also, as discussed caller of unlock should retry if it fails, then should retry also be added here?

Reply:
good catch! I'll add a flag as you said.